File: Test019.py
__author__ = 'zyt'

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param {ListNode} head
    # @param {integer} n
    # @return {ListNode}
    def removeNthFromEnd(self, head, n):
        if not head: return None
        start = end = head
        count = 0
        while count < n:
            end = end.next
            count += 1
        if not end:
            return head.next
        while end.next:
            start = start.next
            end = end.next
        start.next = start.next.next
        return head

File: LeetCode_Java/src/Test019.java
/**
 * Created by zyt on 15/8/21 15:36.
 */
public class Test019 {
    /**
     * Definition for singly-linked list.
     */
    private class ListNode {
        int val;
        ListNode next;
        
        ListNode(int x) {
            val = x;
        }
    }
    
    private class Solution {
        public ListNode removeNthFromEnd(ListNode head, int n) {
            if (head == null) {
                return null;
            }
            ListNode p = head, q = head;
            for (int i = 0; i < n; i++) {
                q = q.next;
            }
            if (q == null) {
                return head.next;
            }
            while (q.next != null) {
                q = q.next;
                p = p.next;
            }
            p.next = p.next.next;
            return head;
        }
    }
}
